[댓글 기능] 이바 미션 제출합니다.

seonghun127 opened the pull request:
안녕하세요. 쏘쏘! 리뷰이 이바입니다.

다른 때보다 미션 수행 기간이 짧았던 만큼 만족스럽지 않은 상태에서 PR을 보내게됐습니다.

이전과 마찬가지로 테스트 코드 작성이 정말 어려웠습니다.
상속, Mockito 등을 사용하면 어떠냐는 피드백을 받았지만 적용하는데 실패하여 
결국 저만의 방법(?)으로 테스트 코드를 작성했습니다.

현재 테스트 코드 전체를 돌리면 무수히 깨지는 상황이고 Test 클래스 별로 테스트 코드가 바르게 돌아가게끔까지
코드를 짰습니다. 이렇게 PR을 보내고 리펙토링을 진행하려 합니다.

부족한 부분이 많을테니 거침없이 지적해주시면 감사하겠습니다.

seonghun127 committed:
setting: mission complete

seonghun127 committed:
docs: README.md

1. todo 목록 정리

seonghun127 committed:
feat: build.gradle

1. devtools 의존성 추가

seonghun127 committed:
feat: ArticleController Class

1. Article 도메인 객체 구현
2. Test 추가
3. ArticleRepository 저장기능 구현

seonghun127 committed:
feat: layout

1. build.gradle 의존성 추가
2. 공통 레이아웃 추출 및 적용

seonghun127 committed:
feat: ArticleDto Class

1. Create 용도의 ArticleDto 구현
2. ArticleDto를 사용하도록 Controller와 Repository 변경

seonghun127 committed:
refactor: Article class

1. ArticleDto를 사용하기 위한 생성자 추가
2. id 자동 증가를 위한 static field 추가

seonghun127 committed:
feat: ArticleController class

1. 조회 기능 구현

seonghun127 committed:
style: css 및 templates 파일 코드 포메팅

seonghun127 committed:
fix: app.css

1. 파일 오류 수정

seonghun127 committed:
test: ArticleRepository class

1. 조회 및 전체조회 테스트

seonghun127 committed:
fix: layout

1. thymeleaf-layout 버전에 따른 오류 수정

seonghun127 committed:
fix: ArticleController class

1. createArticle 리다이렉팅 주소 수정

seonghun127 committed:
feat: ArticleController class

1. 게시물 수정 기능 구현
2. ArticleDto.Update 생성
3. ArticleRepository에 update 기능 추가
4. update를 위해 Article 도메인 객체에 생성자 추가

seonghun127 committed:
feat: article-edit.html

1. 생성 화면과 수정 화면에 따른 form 경로 및 메서드 조정
2. article.html에 하드코딩된 url 변경

seonghun127 committed:
feat: ArticleController class

1. 삭제 기능 구현
2. ArticleRepository delete 구현

seonghun127 committed:
feat: article.html

1. delete 버튼 구현

seonghun127 committed:
refactor: layout

1. header 공통부분 추출

seonghun127 committed:
fix: ArticleRepository class

1. 수정시 새 게시물이 추가되는 오류 수정

seonghun127 committed:
style: 전체 코드 포매팅

seonghun127 committed:
fix: article-edit.html

1. 수정 페이지에서 글 내용이 나오지 않는 오류 수정

seonghun127 committed:
fix: article.html

1. 예시로 설정된 게시글 내용 삭제
2. 작성된 게시글 내용이 나오게끔 수정

seonghun127 committed:
refactor: custom head 적용 방법 변경

seonghun127 committed:
refactor: fragment 폴더 생성

seonghun127 committed:
refactor: layout.html

1. main-head를 담는 th:replace 태그 추가
2. main-body를 담는 th:replace 태그 추가

seonghun127 committed:
refactor: ArticleController class

1. 필드 의존 주입에서 생성자 의존 주입으로 변경

seonghun127 committed:
refactor: NotFoundArticleException class 추가

seonghun127 committed:
refactor: ArticleRepository class

1. find() 메서드의 이름 수정 : findById()

seonghun127 committed:
style: 코드 포매팅

seonghun127 committed:
refactor: Article class

1. set 메서드 제거

seonghun127 committed:
fix: ArticleRepositoryTest 클래스에서 깨지는 테스트 코드 수정

seonghun127 committed:
refactor: ArticleController class

1. @AllArgsContructor 를 @RequiredArgsContructor 어노테이션으로 변경
2. articleRepository 필드를 final 변경
- 불변셩을 유지하기 위해

seonghun127 committed:
feat: Article class

1. 커스터마이즈한 Builder 패턴 추가

seonghun127 committed:
refactor: Article class에 lombok @Builder 적용

seonghun127 committed:
feat: ArticleController class

1. 게시물 추가 테스트 코드 구현

seonghun127 committed:
style: ArticleRepositoryTests class 이름 변경

seonghun127 committed:
fix: Article class

1. updateArticle() 메서드 추가

seonghun127 committed:
refactor: 중복 head 부분 제거

seonghun127 committed:
feat: ArticleController class

1. modelMapper 의존성 주입
2. modelMapper 클래스를 빈으로 등록
3. ArticleDto 클래스에 Response 중첩클래스 추가
4. ArticleController에서 article(s) 를 반환하는 모든 API의 반환 형태를 ArticleDto.Response 타입으로 변경

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: ArticleController class

1. updateArticle() 메서드 리턴 방식 변경
- 리다이렉트에서 포워딩 방식으로 변경한다
2. 그에 따른 인수 테스트 방식 수정
- 수정된 값을 확인하는 테스트로 수정한다

seonghun127 committed:
style: 주석 제거

seonghun127 committed:
refactor: ArticleRespository class

1. save() 메서드 리턴 타입 void로 변경
2. update() 메서드 리턴 타입 void로 변경

seonghun127 committed:
refactor: ArticleController class

1. createArticle() 메서드 리턴 타입 변경
- 리다이렉트 방식에서 포워딩 방식으로 변경한다

seonghun127 committed:
refactor: ArticleControllerTests class

1. 지속 테스트 가능한 코드로 리펙토링

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
fix: ArticleController class

1. createArticle(), updateArticle() api 리턴 타입 변경
- 포워드 방식에서 리다이렉트 방식으로 변경한다

2. ArticleRepository 클래스에서 save(), update() 메서드의 리턴 타입 변경
- void에서 long으로 변경한다

3. 테스트 코드 수정

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
fix: Spring Data Jpa 사용

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
docs: h2 기본 설정

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
feat: 회원가입 요청 페이지 이동 및 저장

1. 회원 가입 페이지 이동 api(renderSignUpPage())를 구현한다
2. 회원 정보를 저장하는 api(createUser())를 구현한다

seonghun127 committed:
merge local branch into fork repository branch

seonghun127 committed:
feat: 회원등록 유효성 검사

seonghun127 committed:
fix: 회원가입 api response 방식 변경

1. createUser() api의 리턴 방식을 리다이렉트로 변경한다
- 회원가입이 성공하면 "/login"으로 리다이렉트 된다
- 회원가입이 실패하면 "/signup"으로 리다이렉트 된다

seonghun127 committed:
feat: 회원 조회 기능 구현

1. 회원을 조회하는 api(readUsers())를 구현한다
2. service 계층에서 dto와 domain 객체의 mapping 처리한다

seonghun127 committed:
fix: 페이지 링크 수정

seonghun127 committed:
feat: 서버에 예외처리 추가

1. email, password, name에 형식 조건을 추가한다
2. ControllerAdvice로 exception handler를 관리한다

seonghun127 committed:
feat: UserControllerTest class

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
refactor: 서버에서 예외처리 리팩토링

1. 커스텀 예외를 추가한다
- 중복, 형식(이메일, 이름, 비밀번호)
2. ControllerAdvice와 ExceptionHandler를 적용한다
3. Response Status를 설정한다

seonghun127 committed:
feat: 로그인, 로그아웃 기능 구현

1. Login 전용 DTO를 추가한다 (UserDTO.Login)
2. 로그인 성공 시 Session을 생성한다
3. 로그인 유무에 따라 페이지를 분기 이동한다

seonghun127 committed:
fix: 페이지 요청 방식 수정

1. get 요청 시 화면, url 일관성 주기 위해 redirect방식으로 변경한다

seonghun127 committed:
docs: 기능 구현 목록 정리

seonghun127 committed:
feat: MyPage 이동 기능 구현

1. 마이페이지로 이동하는 api(renderMYPage())를 구현한다
2. 로그인 유무에 따라 이동 페이지를 결정한다

seonghun127 committed:
fix: 로그인이 안되있을 시 로그아웃 기능 수정

1. session 정보가 존재할때만 로그아웃을 수행한다

seonghun127 committed:
docs: 기능 목룩 구현 정리

seonghun127 committed:
feat: 회원 정보 수정 기능 추가

1. 회원 정보 수정 api(updateUser())를 추가한다
2. My Page로 이동하는 api(renderMypage())를 수정한다
- session 정보를 확인하는 로직을 삭제한다
3. Edit My Page로 이동하는 api(renderEditMypage())를 추가한다
- renderMypage()에서 처리하던 session 관리 로직을 추가한다
4. UserDto.Reponse 클래스에 id 필드를 추가한다
5. Service 계층에 update() 메서드를 추가한다

seonghun127 committed:
fix: 회원가입, 회원수정 시 예외 처리 코드 분류

1. InvalidSignUpFormException을 구현한다
2. InvalidEditFormException을 구현한다

seonghun127 committed:
feat: 회원 탈퇴 기능 구현

1. 회원 정보를 삭제하는 api(deleteUser()) 추가한다
2. 테스트 코드 @AfterEach에 삭제 api를 호출하여 전체 테스트가 가능하게 한다

seonghun127 committed:
feat: UserService 테스트 코드 추가

seonghun127 committed:
fix: UserService 테스트 코드 수정

seonghun127 committed:
refactor: ArticleService 계층 분리

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
docs: data.sql 추가

seonghun127 committed:
fix: 테스트 코드 오류 수정

1. data.sql 파일 삭제

seonghun127 committed:
refactor: 도메인 @Builder 어노테이션 적용 위치 수정

1. class 전체에 적용하던 @Builder 어노테이션을 생성자로 변경한다
- 생성자 파라미터에 대해서만 빌더 패턴을 적용하기 위해

seonghun127 committed:
refactor: 회원 삭제 시 logout api 호출로 로직 변경

seonghun127 committed:
refactor: 도메인별 패키지 분리

seonghun127 committed:
refactor: 로그인 여부에 따른 이동경로 분기 설정

1. 화면 우측 상단에 이름 / Login 버튼에 대해 로그인 여부에 따라
이동하는 경로를 달리 지정한다
- 로그인 시 '/', 아닐 시 '/login'으로 이동한다

seonghun127 committed:
feat: Interceptor 적용

seonghun127 committed:
feat: Argument Resolver 구현

seonghun127 committed:
refactor: session 정보 삭제 방법 수정

1. invalidate() 메서드를 removeAttribute() 메서드로 변경한다

seonghun127 committed:
refactor: dto 클래스의 중첩 클래스 이름을 명사형으로 수정

seonghun127 committed:
style: 코드 포멧팅

seonghun127 committed:
feat: UserDto.Login 클래스에 validate 추가

1. 이메일과 비밀번호의 형식을 체크한다
2. InvalidLoginFormException을 추가한다

seonghun127 committed:
feat: User 클래스의 필드 validation 추가

seonghun127 committed:
refactor: update 수정

1. 변경할 객체를 생성하고 save()를 다시 호출했던 방식에서
변경할 객체에 update() 메세지를 보내어 수정되는
사항이 영속적으로 저장되는 방식으로 변경한다

seonghun127 committed:
style: import 패키지 변경

seonghun127 committed:
refactor: 로그인시 비밀번호 일치 검사 로직 변경

1. User 클래스에 checkPassword() 메서드를 추가한다
2. User 객체에게 checkPassword() 메세지를 보내 비밀번호를 확인한다

seonghun127 committed:
refactor: 테스트 시 회원가입과 로그인 로직을 따로 추출

1. test code에서 setUp()에서 처리했던 회원가입, 로그인하는 로직을
UserUtilForTest 클래스로 추출한다

seonghun127 committed:
feat: Attribute Converter 구현

1. User 도메인 필드의 타입에 대한 래핑 클래스를 생성한다
2. 서버에선 User 도메인 필드 타입을 모드 래핑 클래스로 적용한다
3. DB에 저장할 시 String 타입으로 저장해주는 래핑 클래스에 대한 Converter를 생성한다

seonghun127 committed:
refactor: Name 클래스의 updateName 메서드 수정

1. VO인 Name 클래스의 불변을 위해 updateName() 메서드 로직을
변경될 필드 값에 대해 새로운 Name 객체로 반환하는 구조로 변경한다

seonghun127 committed:
refactor: VO 클래스 필드에 final 추가

seonghun127 committed:
refactor: domain 패키지 안에 vo 패키지 추가

seonghun127 committed:
refactor: 회원 정보 수정 성공 테스트 리펙토링

1. 정보 수정 요청 후 리다이렉트 되는 페이지에서 수정된 정보가
잘 반영됐는지 확인하는 로직을 추가한다

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
Merge remote-tracking branch 'upstream/seonghun127' into seonghun127

seonghun127 committed:
docs: 기능 목록 구현 업데이트

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 다대일 매핑

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 수정, 삭제 시 회원 정보 확인

seonghun127 committed:
feat: 게시글 회원 정보 연결

1. 프론트에서 프로필을 회원 정보에 맞게 변경

seonghun127 committed:
feat: 댓글 생성 기능 추가

1. User 도메인과 Article 도메인을 연결한다

seonghun127 committed:
refactor: Article 클래스의 메서드명 변경

1. matchAuthorId() 에서 notMatchAuthorId()로 변경한다

seonghun127 committed:
feat: 댓글 조회, 수정, 삭제 기능 추가

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: @Data 어노테이션 추가

seonghun127 committed:
feat: 댓글 생성, 조회, 삭제 기능 구현

seonghun127 committed:
feat: 댓글 수정 기능 구현

seonghun127 committed:
feat: 테스트 코드 구현

seonghun127 committed:
feat: 오류 메세지 출력

ssosso reviewed (changes requested):
안녕하세요, 이바~ 쏘쏘입니다:) (저와의 코드리뷰에서는 '님' 어미를 사용하지 않으셔도 됩니다)
만족스럽지 않으셨다고 했는데 정갈한 코드네요. 고생하셨습니다.
전체 테스트 코드를 실행하면 실패하는 부분은 확인해서 반드시 고치면 좋을 것 같습니다.
피드백 드린 부분 외에 궁금한 부분이 있으시면 말씀주세요:)

seonghun127 committed:
refactor: RedirectView 클래스 사용

1. ArticleController에서 redirect되는 api의 리턴 타입을 RedirectView로 변경한다

seonghun127 committed:
refactor: forward 되는 api의 리턴 값에서 '/' 제거

seonghun127 committed:
refactor: 세션 정보 확인 로직 수정

1. UserSession의 id 값과 uri 쿼리를 비교하는 로직으로 변경한다

seonghun127 committed:
refactor: 테스트 코드 중복 제거

seonghun127 committed:
refactor: ServiceTest 오류 수정

seonghun127 committed:
refactor: 예외 클래스에 정확한 로그 기록을 위해 생성자에 인자 추가

1. User, Article, Comment를 찾을 수 없는 예외 발생 시 해당 객체의 id 값을
예외 클래스에 넘겨주어 id까지 예외 메세지로 출력한다

seonghun127 committed:
refactor: dto 분리

1. 도메인별 DTO 클래스에 중첩 클래스를 사용했던 로직을
따로 분리한다

seonghun127 committed:
refactor: @EqualsAndHashCode 어노테이션 속성 설정

1. Immutable field인 id값으로만 equals, hashcode를 구현하게 한다
- 모든 필드 비교시 발생하는 문제를 방지하기 위함

seonghun127 committed:
refactor: Article, Comment Service 레이어의 중복 코드 제거

1. update, delete 메서드 수행 시 본인인지 확인하는 로직을
checkAuthority() 메서드로 분리한다

seonghun127 committed:
refactor: fragment/header.html을 index.html에서도 적용

seonghun127 committed:
refactor: User Service에서 deleteById() 메서드에 try-catch 문 추가

1. JPA 기본 메서드인 deleteById() 메서드에서 던저지는 오류를 잡고
커스텀으로 만든 NotFoundUserException을 던진다

seonghun127 committed:
style: 코드 포맷팅

seonghun127 commented:
쏘쏘 안녕하세요. 리뷰이 이바입니다.

유용한 정보와 함께 피드백 주신 점에 대해 진심으로 감사합니다.
쏘쏘 언급한 내용을 최대한 반영하려고 노력했습니다.

하지만 테스트 코드에 대해서는 아직 부족한 부분이 너무 많네요.
```
webTestClient.~ .exchange()
```
위 코드의 중복을 http method에 따라 만들어 놓고 재사용하는 방식까진 적용했지만
전체 테스트 코드를 돌릴 시 테스트 코드가 깨지는 문제는 아직 해결하지 못했습니다.

인수 테스트 시 각각의 도메인 객체에 접근하기 위한 id 값 추출에 어려움을 겪고 있습니다. 임시방편으로 
`public static long userId = 1` 과 같이 전역 변수를 두어 도메인 생성시 마다 1씩 증가 시켜주는 방식을 통해
오류 코드를 해결했지만 전체 테스트 코드 수행시, 다른 클래스에서 생성되어 id가 몇인지 모르는 상태에서
무작정 id를 1로 접근하려하니 테스트 코드가 무수히 깨져버리고 있습니다.

물론 인수 테스트 코드에서 Repository를 필드로 두어 현재 id의 값이 몇인지 조회해보는 방법도 있겠지만
인수 테스트에 Repository에 의존하기 싫어 이 방법을 피하고 다른 방법을 계속 고민해보고 있습니다.
테스트 코드가 정말 어렵네요 ㅠ

부족한 부분이 많다고 생각하기에 다시 한번 많은 지적 부탁드리겠습니다. 감사합니다.

ssosso reviewed (approved):
이바~ 정말 많은 고민을 하셨네요:) 보기 좋습니다~
테스트 코드 전체 실행 시 문제가 되는 부분에 대해서 말씀주셨는데,
전반적인 테스트 코드에서 사용할 수 있는 임의 데이터 삽입 메서드를 만들어 두고 이를 통해 테스트 전에 임의 데이터 생성하고 그에 대한 id를 받아와 처리하는 것이 좋습니다.
그렇지 않으면 병렬 테스트 처리 과정에서 당연히 실패할 수 밖에 없어요~
그 외 댓글로 질문주신 내용도 각각 답변 달아두었습니다:)~
모두 반영했다는 생각이 드시면 말씀주세요 머지하겠습니다.

seonghun127 committed:
refactor: @RequiredArgsConstructor 어노테이션 삭제

seonghun127 committed:
refactor: User 도메인 클래스에 @EqualsAndHashCode 어노테이션 삭제

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: void 메서드에 반환값이 있도록 변경

seonghun127 committed:
style: 코드 포맷팅

seonghun127 committed:
refactor: 테스트 코드에 data.sql와 mokito 적용

seonghun127 commented:
안녕하세요 쏘쏘! 리뷰이 이바입니다.

테스트 코드를 전폭 수정했습니다. .sql 파일로 데이터 주입을 임의로 하여 인수 테스트를 진행하고 mockito를 사용하여 service 레이어 테스트 코드를 수정했습니다. 이제는 전체 실행 시 전부 통과합니다 👍 

반환값에 대해 고민하면서 update에 대해서는 반환값을 두었지만 아직 delete를 할 때 적절하게 반환값을 두고 처리하는 방법에 대해 마음에 드는 로직을 구현하지 못해 반영하진 못했습니다 ㅠㅠ

그렇지만 반환값을 두는 중요성을 알고 향후 프로젝트에는 적극적으로 쏘쏘의 피드백을 따르려고 합니다 ㅎㅎㅎ
수정사항에 대해 부족한 부분이 있다면 거침없이 지적해주시면 감사하겠습니다! (물론 머지도 감사하겠습니다 ㅎ_ㅎ)

ssosso reviewed (approved):
좋네요:) 열정 이바ㅋㅋㅋ  정말 고생많으셨습니다.
제가 드린 피드백에서 혹시 더 좋은 피드백을 받거나 공유해주실 내용이 있다면 언제든지 공유해주세요:)
화이팅~
